expr/func_aspercent.go

```diff
@@ -0,0 +1,254 @@
+	metaSeries := groupSeriesByKey(series, s.nodes, &keys)
+	// The totals series for each key
+	var totalSeries map[string]models.Series
+
+	// calculate the sum
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalSeries = getTotalSeries(metaSeries)
+		// calculate sum of totals series
+	} else if totals != nil {
+		totalSeriesLists := groupSeriesByKey(totals, s.nodes, &keys)
+		totalSeries = getTotalSeries(totalSeriesLists)
+	}
+
+	for key := range keys {
+		// No input series for a corresponding total series
+		if _, ok := metaSeries[key]; !ok {
+			serie2 := totalSeries[key].Copy(pointSlicePool.Get().([]schema.Point))
+			serie2.QueryPatt = fmt.Sprintf("asPercent(MISSING,%s)", serie2.QueryPatt)
+			serie2.Target = fmt.Sprintf("asPercent(MISSING,%s)", serie2.Target)
+			serie2.Tags = map[string]string{"name": serie2.Target}
+			for i := range serie2.Datapoints {
+				serie2.Datapoints[i].Val = math.NaN()
+			}
+			outSeries = append(outSeries, serie2)
+			continue
+		}
+
+		for _, serie1 := range metaSeries[key] {
+			serie1 = serie1.Copy(pointSlicePool.Get().([]schema.Point))
+			// No total series for a corresponding input series
+			if _, ok := totalSeries[key]; !ok {
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,MISSING)", serie1.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,MISSING)", serie1.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = math.NaN()
+				}
+			} else {
+				serie2 := totalSeries[key]
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+				}
+			}
+			outSeries = append(outSeries, serie1)
+		}
+	}
+	return outSeries, nil
+}
+
+func (s *FuncAsPercent) execWithoutNodes(series, totals []models.Series) ([]models.Series, error) {
+	var outSeries []models.Series
+	var totalsSerie models.Series
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalsSerie = sumSeries(series)
+		if len(series) == 1 {
+			totalsSerie.Target = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.QueryPatt = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.Tags = map[string]string{"name": totalsSerie.Target}
+		}
+	} else if totals != nil {
+		if len(totals) == 1 {
+			totalsSerie = totals[0]
+		} else if len(totals) == len(series) {
+			// Sorted to match the input series with the total series based on Target.
+			// Mimicks Graphite's implementation
+			sort.Slice(series, func(i, j int) bool {
+				return series[i].Target < series[j].Target
+			})
+			sort.Slice(totals, func(i, j int) bool {
+				return totals[i].Target < totals[j].Target
+			})
+			for i, serie1 := range series {
```

Comment:
lots of duplication here wrt the similar code block further down.
can't we do all the if/else stuff here to just set up the right totals and series variables,
and then do the same processing at the end irrespective of which scenario it was?